core/iwasm/common/wasm_shared_memory.c

```diff
@@ -435,40 +444,48 @@ wasm_runtime_atomic_wait(WASMModuleInstanceCommon *module, void *address,
     if (no_wait) {
         return 1;
     }
-    else {
-        bh_list_status ret;
 
-        if (!(wait_node = wasm_runtime_malloc(sizeof(AtomicWaitNode)))) {
-            wasm_runtime_set_exception(module, "failed to create wait node");
-            return -1;
-        }
-        memset(wait_node, 0, sizeof(AtomicWaitNode));
+    if (!(wait_node = wasm_runtime_malloc(sizeof(AtomicWaitNode)))) {
+        wasm_runtime_set_exception(module, "failed to create wait node");
+        return -1;
+    }
+    memset(wait_node, 0, sizeof(AtomicWaitNode));
 
-        if (0 != os_mutex_init(&wait_node->wait_lock)) {
-            wasm_runtime_free(wait_node);
-            return -1;
-        }
+    if (0 != os_mutex_init(&wait_node->wait_lock)) {
+        wasm_runtime_free(wait_node);
+        return -1;
+    }
 
-        if (0 != os_cond_init(&wait_node->wait_cond)) {
-            os_mutex_destroy(&wait_node->wait_lock);
-            wasm_runtime_free(wait_node);
-            return -1;
-        }
+    if (0 != os_cond_init(&wait_node->wait_cond)) {
+        os_mutex_destroy(&wait_node->wait_lock);
+        wasm_runtime_free(wait_node);
+        return -1;
+    }
 
-        wait_node->status = S_WAITING;
-        os_mutex_lock(&wait_info->wait_list_lock);
-        ret = bh_list_insert(wait_info->wait_list, wait_node);
-        os_mutex_unlock(&wait_info->wait_list_lock);
-        bh_assert(ret == BH_LIST_SUCCESS);
-        (void)ret;
+    wait_node->status = S_WAITING;
+
+    /* acquire the wait info, create new one if not exists */
+    wait_info = acquire_wait_info(address, true, wait_node);
+
+    if (!wait_info) {
+        os_mutex_destroy(&wait_node->wait_lock);
+        wasm_runtime_free(wait_node);
+        wasm_runtime_set_exception(module, "failed to acquire wait_info");
+        return -1;
     }
 
     /* condition wait start */
     os_mutex_lock(&wait_node->wait_lock);
 
-    os_cond_reltimedwait(&wait_node->wait_cond, &wait_node->wait_lock,
-                         timeout < 0 ? BHT_WAIT_FOREVER
-                                     : (uint64)timeout / 1000);
+#if WASM_ENABLE_THREAD_MGR != 0
+    if (!wasm_cluster_is_thread_terminated(exec_env)) {
```

Comment:
Suggest to also re-read the value in address again, which may have changed after we read it last time:
```C
    os_mutex_lock(&node->shared_mem_lock);
    no_wait = (!wait64 && *(uint32 *)address != (uint32)expect)
              || (wait64 && *(uint64 *)address != expect);
    os_mutex_unlock(&node->shared_mem_lock);
    if (!no_wait
#if WASM_ENABLE_THREAD_MGR != 0
        && !wasm_cluster_is_thread_terminated(exec_env)
#endif
       ) {
         os_cond_reltimedwait(&wait_node->wait_cond, &wait_node->wait_lock,
                             timeout < 0 ? BHT_WAIT_FOREVER
                                         : (uint64)timeout / 1000);
      }
```
And in the end of this function:
`return no_wait ? 1 : is_timeout ? 2 : 0;`

Reply:
It's done in L440 already, since `address` and `expect` don't change in this function then there's no need to call it here.

Reply:
Though address don't change, the value of `*address` may be changed by other thread after L440?

Reply:
Okay I will add it, I believe I don't understand yet how it might be used and purpose of `*address` in general. Can you give an example why/how it can be changed?

Reply:
Another thread may store data into the address with atomic.i32/i64.store opcodes after this thread read it in L442:

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_interp_classic.c#L3550-L3623

https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_interp_classic.c#L3577

Reply:
Fixed